cvs/objects/sectors/source/value_factor_calculator.cpp

```diff
@@ -0,0 +1,173 @@
+#include "marketplace/include/marketplace.h"
+
+using namespace std;
+
+/*!
+ * \brief Constructor.
+ */
+ValueFactorCalculator::ValueFactorCalculator()
+{
+    mValueFactorIntercept = 1.0;
+    mValueFactorSlope = 0.0;
+}
+
+// Documentation is inherited.
+ValueFactorCalculator* ValueFactorCalculator::clone() const {
+    ValueFactorCalculator* clone = new ValueFactorCalculator();
+    clone->mValueFactorIntercept = mValueFactorIntercept;
+    clone->mValueFactorSlope = mValueFactorSlope;
+    
+    return clone;
+}
+
+// Documentation is inherited.
+bool ValueFactorCalculator::isSameType( const std::string& aType ) const {
+    return aType == getXMLNameStatic();
+}
+
+// Documentation is inherited.
+const string& ValueFactorCalculator::getName() const {
+    return getXMLNameStatic();
+}
+
+/*! \brief Get the XML node name in static form for comparison when parsing XML.
+*
+* This public function accesses the private constant string, XML_NAME. This way
+* the tag is always consistent for both read-in and output and can be easily
+* changed. The "==" operator that is used when parsing, required this second
+* function to return static.
+* \note A function cannot be static and virtual.
+* \return The constant XML_NAME as a static.
+*/
+const string& ValueFactorCalculator::getXMLNameStatic() {
+    const static string XML_NAME = "value-factor-calculator";
+    return XML_NAME;
+}
+
+const string& ValueFactorCalculator::getXMLName() const {
+    return getXMLNameStatic();
+}
+
+// Documentation is inherited.
+void ValueFactorCalculator::toDebugXML( const int aPeriod, ostream& aOut, Tabs* aTabs ) const {
+    XMLWriteOpeningTag( getXMLNameStatic(), aOut, aTabs );
+    XMLWriteElement(mValueFactorIntercept, "ValueFactorIntercept", aOut, aTabs );
+    XMLWriteElement(mValueFactorSlope, "ValueFactorSlope", aOut, aTabs );
+    XMLWriteClosingTag( getXMLNameStatic(), aOut, aTabs );
+}
+
+// Documentation is inherited.
+void ValueFactorCalculator::initCalc( const IInfo* aTechInfo ) {
+    // No information needs to be passed in
+}
+
+double ValueFactorCalculator::getMarginalBackupCapacity(const string& aSector,
+    const string& aElectricSector,
+    const string& aResource,
+    const string& aRegion,
+    const double aTechCapacityFactor,
+    const double aReserveMargin,
+    const double aAverageGridCapacityFactor,
+    const int aPeriod) const
+{
+    // This is a placeholder function so we can use the IBackupCalculator class
+    // For now, it will do nothing (and should not be called)
+
+    return 0.0;
+}
+
+double ValueFactorCalculator::getAverageBackupCapacity(const string& aSector,
+    const string& aElectricSector,
+    const string& aResource,
+    const string& aRegion,
+    const double aTechCapacityFactor,
+    const double aReserveMargin,
+    const double aAverageGridCapacityFactor,
+    const int aPeriod) const
+{
+    // This is a placeholder function so we can use the IBackupCalculator class
```

Comment:
okay, so my rust is showing.  Is it because it is pure virtual and needs to be defined?  could it instead be removed from the ibackup_calculator, or do you think it might be useful for something else so should be left in?

Reply:
i guess we need to see how to deal with Page's comments on maintaining the old ways as an option perhaps

Reply:
I don't think we want to support multiple ways of representing intermittency; pretty much every time we have ever tried to maintain multiple ways of doing something, the “non-default” option only gets maintained for a few months. We don't want to require ourselves to include multiple representations of intermittency in our standardized testing suite.
The issue that I have raised with this proposal is that right now, in high renewable scenarios, we see lots of single-cycle combustion turbines, whether powered by hydrogen or natural gas. Almost all of this capacity is in the backup electricity sectors. Perhaps coincidentally, the total capacity and total electricity output of these combustion turbines is reasonably consistent with ReEDS; high renewable scenarios in the USA tend to see between 500 and 1,000 gigawatts of combustion turbine capacity. In my opinion that technology should not be turned off in this revised approach; there should still be some demand for the services provided by combustion turbines, and their costs should be passed forward to electricity consumers.

Reply:
Hi @pkyle, thanks for your review comments.

One thing to note is that the new PLCOE method does allow us to calculate a total integration cost per region / technology / model period by taking the difference of PLCOE and LCOE.  This integration cost reflects a variety of technologies that provide system flexibility and support VRE integration (e.g., storage, transmission, and firm capacity resources), and could be used to post-calculate implied combustion turbine capacity or some mix of combustion turbines, storage, and transmission.  It’s not obvious that single-cycle combustion turbines are the primary system integration option supporting VRE.  Per the table below for the ReEDS 2023 Standard Scenarios Mid Case, the growth in battery capacity from 2010 through 2050 is twice that of CT capacity. Doing a robust competition between Storage, CT, and other integration options is beyond the scope of this CMP, but we have done some preliminary thinking on a hybrid approach that combines PLCOE-based competition with dynamic calculation of storage/firm capacity.  This is a good opportunity for future model development.

  | 2010 | 2020 | 2030 | 2040 | 2050
-- | -- | -- | -- | -- | --
Wind Capacity (GW) | 40 | 118 | 441 | 641 | 762
PV Capacity (GW) | 1 | 72 | 352 | 753 | 1095
VRE Capacity (GW) | 41 | 190 | 793 | 1393 | 1857
Battery Capacity (GW) | 0 | 0 | 50 | 197 | 369
Gas-CT Capacity (GW) | 117 | 128 | 123 | 173 | 300
H2-CT Capacity (GW) | 0 | 0 | 0 | 11 | 11
CT Capacity (GW) | 117 | 128 | 123 | 183 | 311
Battery-VRE Capacity Ratio | 0.00 | 0.00 | 0.06 | 0.14 | 0.20
CT-VRE Capacity Ratio | 2.87 | 0.67 | 0.16 | 0.13 | 0.17

*Table from ReEDS 2023 Standard Scenarios Mid Case*

RE: supporting multiple ways of representing intermittency, @mbins talked to @pralitp and worked out an approach for maintaining the previous approach in code that should be reasonably robust.  You’re right that the “non-default” option will be more susceptible to lapses in maintenance.  While we feel that the PLCOE approach proposed in this CMP does a better job of representing VRE competitiveness (and is more transparently parameterized), we acknowledge that removing the ability to estimate firm backup capacity (and associated energy use / emissions) does remove a capability that’s valuable for some projects.  Thus, we agree with @mwisepnnl's suggestion to maintain the ability to use the current approach for the subset of projects for which estimating firm backup capacity is important.

-@mmowers and @mbins

Reply:
Thanks for the context, and comparison with the published ReEDS run. As I noted in my initial comment, what is most important here is how this proposal impacts the system level demands for natural gas and hydrogen. Levels of deployment of hydrogen combustion turbines in ReEDS are sensitive to the scenario chosen, through the ratio of total combustion turbine capacity to renewable energy capacity is presumably more stable. And from GCAM’s perspective, the battery option would be a system cost and electricity loss that is passed on to end use consumers of electricity, similar to T&D in general.
I suppose for my own purposes, if I can continue to use the backup electricity representation that we have right now for the next year or two, while we coalesce on a strategy for how to represent short term capacity and generation, which would value combustion turbines appropriately, that does suffice. But longer term I would not want to see the demand for hydrogen combustion turbines go to 0 simply because of a different modeling convention that we adopted. Right now this is one of the major demands for hydrogen we see in GCAM, system wide in net-zero scenarios, and the deployment levels are consistent with the current literature.

Reply:
> okay, so my rust is showing. Is it because it is pure virtual and needs to be defined? could it instead be removed from the ibackup_calculator, or do you think it might be useful for something else so should be left in?

Marshall - I think you are correct.  We may have been able to remove the getMarginalBackupCapacity and getAverageBackupCapacity from the IBackupCalculator class (from which the ValueFactorCalculator class is derived).  But now that we've reintroduced the CapacityLimitBackupCalculator class, which uses those methods, I think we need to leave them in the IBackupCalculator class and thus leave empty implementations of those methods in the ValueFactorCalculator class.